### PR TITLE
feat(test): Add code coverage for karma runs

### PR DESCRIPTION
**Pull request by delasteve**

### CODE REVIEW

**Comment by delasteve** [comment]
I think my accidental closing of the issue caused this to not build on AppVeyor. Is there a way rerun the build manually?

File: addon/ng2/blueprints/ng2/files/package.json [hunk 1]
```diff
@@ -42,7 +42,9 @@
     "jasmine-spec-reporter": "^2.4.0",
     "karma": "^0.13.15",
     "karma-chrome-launcher": "^0.2.3",
+    "karma-coverage": "^1.0.0",
     "karma-jasmine": "^0.3.8",
+    "karma-remap-istanbul": "git+https://github.com/marines/karma-remap-istanbul.git#patch-1",
```
Comment (hansl) [inline addon/ng2/blueprints/ng2/files/package.json]: I'm sorry but I can't allow every CLI dependencies to be reliant on a ref on someone's GitHub.

Comment (delasteve) [inline addon/ng2/blueprints/ng2/files/package.json]: I agree. So, there are two options:

1. We wait for the dev to implement the PR for `karma-remap-istanbul`
2. We/I/Someone creates `karma-remap-istanbul2` with the fix.

Suggestions?

**Comment by delasteve** [comment]
@hansl, I got in contact with @marcules, the maintainer for `karma-remap-istanbul` and was able to release a new version (`0.0.6`). This should take care of your concern (and mine, too).

The only concern I have left is the fact everything is considered to be under `__root__`. This is an issue with `remap-istanbul` (reported here: https://github.com/SitePen/remap-istanbul/issues/50). 

Let me know if there is anything else you would like me fix! Thanks.

**Comment by delasteve** [comment]
Found another concern, maybe someone can help. I am getting `Not all files specified in sources could be found, continue with partial remapping.` Has anyone experienced this or know what I am doing wrong?

**Comment by filipesilva** [comment]
@delasteve can you tell me how to repro that error?

**Comment by delasteve** [comment]
@filipesilva: it's an issue with `karma-remap-istanbul`. I am hesitant to change things too much with no tests. I don't have time to maintain the project (a permission I was given by the owner due to his lack of time to merge a PR).

I just released a version that will fix the timeout issue that was causing that error line. It was always executing the timeout even if all the files were found.

I just fixed the code to use the new version. It's going through testing now.

**Comment by delasteve** [comment]
@filipesilva would you mind rerunning the build? AppVeyor looks like it had a network issue and while the build completed and tests were the same I had on my local, the TravisCI also failed.

Thanks!

**Comment by filipesilva** [comment]
@delasteve I restarted both builds now.

**Comment by delasteve** [comment]
I will take a look into the failures when I get home in a bit. Curious if it has something to do with my update to the `karma-remap-istanbul` plugin.

**Comment by devCrossNet** [comment]
@delasteve @filipesilva I tried the same approach as you and had the issue that the html report was not generated on the first run. Do you have the same issue? I have another approach that also removes the dependency to karma-remap-istanbul, you can check it here: https://github.com/devCrossNet/angular-cli/tree/coverage. I also made a pull-request to the remap-istanbul project that should solve the "missing folder structure" issue, you can see it here: https://github.com/SitePen/remap-istanbul/pull/65. My approach will work after the pull-request is merged and I update the remap-istanbul dependecy. Let me know what you think.

**Comment by delasteve** [comment]
@filipesilva I took a look locally, and it passes there. I don't know what's going on with CI.

@devCrossNet I like your approach towards the solution. I am not a fan of `karma-remap-istanbul` as it depends on another plugin to run first. That is most likely (although unconfirmed) the cause of your html report woes (coverage reporter wasn't finished yet). 

I wonder if we go the route of a new task, should we have a `ng test --coverage` flag?

**Comment by devCrossNet** [comment]
@delasteve I think coverage should be the default option. but with my solution, it is possible to implement a --coverage flag.

**Comment by delasteve** [comment]
@devCrossNet I agree about it being default. At one point, `--watch` was a thing, but now it's the default. I just want to make sure the discussion happens before this goes in.

Also, I like your solution to this better than using the flakey karma plugin. How should we go about this? I would rather not just lift your code and put my name on the commit. Thoughts?

**Comment by devCrossNet** [comment]
@delasteve I am really new into contributing to projects on GitHub. I have no idea. Maybe you can merge my changes with yours? Then it should be a PR from us both

**Comment by filipesilva** [comment]
@delasteve to be perfectly honest I don't know why travis is failing. The tests all pass and then it seems to hang. Perhaps there is a zombie process created by the new libs that isn't killed?

**Comment by delasteve** [comment]
@filipesilva I almost want to say forget my PR and go with @devCrossNet's solution. It's been running for over 2 hours at this point on Travis. I am not found of the solution `karma-remap-istanbul` puts forth (waiting on another reporter to run), nor do I have want to try and fix it.

@devCrossNet: I am pretty sure the Angular team would like 1 commit per PR. Git only allows 1 author per PR. Personally, I would go ahead and rebase your changes on top of `master`, squash them into 1 commit, and submit a new PR.

**Comment by devCrossNet** [comment]
@delasteve ok, as soon as my changes are merged into remap-istanbul and I finished this whole topic. I will submit a new PR for this feature.

**Comment by devCrossNet** [comment]
@delasteve I updated my repository to use my fork from remap-istanbul. In my opinion, this can be a temporary solution for this feature. But I have another question - should the code coverage also work with ng test --watch=true?

**Comment by delasteve** [comment]
@devCrossNet Got in touch with someone from SitePen. They will be taking a look at your PR in a bit. As for running with `--watch=true`, I would say yes.

**Comment by filipesilva** [comment]
Don't worry too much about the number of commit guys, once it's ready to merge in I'll just squash them myself. 

Regarding the watch, I agree that it should get coverage as well by default. If you implement a `--coverage` flag, then it should correctly interact with `--watch.`, otherwise it can just default to having coverage everywhere.

**Comment by devCrossNet** [comment]
@delasteve @filipesilva thank you very much, guys. I will try to implement the coverage with the watch=true flag. I have no idea how to do that because the karma-server event "run_completed" is triggered before the coverage-final.json is written. maybe I have to set a timeout or something.

**Comment by devCrossNet** [comment]
@delasteve Now it works for ng test and ng test--watch=false. It is really fast and the usage feels good. It would be great if you can do a little review and maybe test it. https://github.com/devCrossNet/angular-cli/commit/a7b3ebe81c4c4b325537d12319c627d73db9222c

**Comment by delasteve** [comment]
@devCrossNet it appears that there is an issue with [your pull request](https://github.com/SitePen/remap-istanbul/pull/65) for `remap-istanbul`. Do you need some help with it?

After that goes in, I definitely will be able to test your changes.

**Comment by devCrossNet** [comment]
@delasteve sorry for the delay. I was sick the last view days. I have to figure out what is the problem with remap-istanbul. Maybe the solution is just to test the values. I will ask him to give a reproducible example. It would be great if you can help me if you know the remap-istanbul project.

**Comment by devCrossNet** [comment]
@delasteve remap-istanbul merged my changes today. after they released the new npm version, I will update my code and make a PR. regards.

**Comment by filipesilva** [comment]
Closed as this PR was made obsolete by https://github.com/angular/angular-cli/pull/1455. Code coverage is in that PR though!

**Comment by Headcult** [comment]
I'm still facing an issue when I implemented Karma-Coverage in my Angular-CLI project. I’m not able to see the correct file names and contents in the HTML report. The coverage is fine but, the file location and name is messed up, it looks like it’s being picked up from the ‘tmp’ folder (tmp/broccoli_type_script_compiler-input_base_path-xrfIqZiR.tmp/0/src/app/). The remapping may have gone wrong. Any workaround or a fix too see the correct file contents within the report. Thanks in advance.

**Comment by filipesilva** [comment]
@Headcult try the webpack preview build, that should be fixed: https://github.com/angular/angular-cli#webpack-preview-release-update

**Pull request closed without merging**